b2sdk/urllib3/_connection.py

```diff
@@ -0,0 +1,238 @@
+    this against AWS services.
+
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._original_response_cls = self.response_class
+        # This variable is set when we receive an early response from the
+        # server. If this value is set to True, any calls to send() are noops.
+        # This value is reset to false every time _send_request is called.
+        # This is to workaround changes in urllib3 2.0 which uses separate
+        # send() calls in request() instead of delegating to endheaders(),
+        # which is where the body is sent in CPython's HTTPConnection.
+        self._response_received = False
+        self._expect_header_set = False
+        self._send_called = False
+
+    def close(self):
+        super().close()
+        # Reset all of our instance state we were tracking.
+        self._response_received = False
+        self._expect_header_set = False
+        self._send_called = False
+        self.response_class = self._original_response_cls
+
+    def request(self, method, url, body=None, headers=None, *args, **kwargs):
+        if headers is None:
+            headers = {}
+        self._response_received = False
+        if headers.get('Expect', b'') in [b'100-continue', '100-continue']:
+            self._expect_header_set = True
+        else:
+            self._expect_header_set = False
+            self.response_class = self._original_response_cls
+        rval = super().request(method, url, body, headers, *args, **kwargs)
+        self._expect_header_set = False
+        return rval
+
+    def _convert_to_bytes(self, mixed_buffer):
+        # Take a list of mixed str/bytes and convert it
+        # all into a single bytestring.
+        # Any str will be encoded as utf-8.
+        bytes_buffer = []
+        for chunk in mixed_buffer:
+            if isinstance(chunk, str):
+                bytes_buffer.append(chunk.encode('utf-8'))
+            else:
+                bytes_buffer.append(chunk)
+        msg = b"\r\n".join(bytes_buffer)
+        return msg
+
+    def _send_output(self, message_body=None, *args, **kwargs):
+        self._buffer.extend((b"", b""))
+        msg = b"\r\n".join(self._buffer)
+        del self._buffer[:]
+        # If msg and message_body are sent in a single send() call,
+        # it will avoid performance problems caused by the interaction
+        # between delayed ack and the Nagle algorithm.
+        if isinstance(message_body, bytes):
+            msg += message_body
+            message_body = None
+        self.send(msg)
+        if self._expect_header_set:
+            # This is our custom behavior.  If the Expect header was
+            # set, it will trigger this custom behavior.
+            logger.debug("Waiting for 100 Continue response.")
+            # Wait for 1 second for the server to send a response.
+            if urllib3.util.wait_for_read(self.sock, 2):
```

Comment:
lol, this is a good case showing why putting an obvious is decremental
it says 1 second, but wait is actually 2 ;p

as to why I'm commenting on this: we may need this to be user-configurable
https://github.com/reef-technologies/b2-sdk-python/pull/252/files#diff-ac825af986ae146afcaa98597de5a9c07399afb45b45edbba3d2b7ab68db124eR26

Please for now just expose it so it is controllable for HTTPAdapterWithContinue user and use `10` as default since that value was chosen with pycurl for some reason.